udp/writer.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2015 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+package udp
+
+import (
+	"bytes"
+	"encoding/binary"
+	"time"
+
+	"github.com/chihaya/chihaya/tracker/models"
+)
+
+// Writer implements the tracker.Writer interface for the UDP protocol.
+type Writer struct {
+	buf *bytes.Buffer
+
+	connectionID  []byte
+	transactionID []byte
+}
+
+// WriteError writes the failure reason as a null-terminated string.
+func (w *Writer) WriteError(err error) error {
+	w.writeHeader(errorActionID)
+	w.buf.WriteString(err.Error())
+	w.buf.WriteRune('\000')
+	return nil
+}
+
+// WriteAnnounce encodes an announce response according to the UDP spec.
+func (w *Writer) WriteAnnounce(res *models.AnnounceResponse) error {
+	w.writeHeader(announceActionID)
```

Comment:
Is it referred to a header anywhere else outside this project? I kind of like each method almost literally being the definition of the spec. For example, the spec says for scrape response:

```
Offset      Size            Name            Value
0           32-bit integer  action          2 // scrape
4           32-bit integer  transaction_id
8 + 12 * n  32-bit integer  seeders
12 + 12 * n 32-bit integer  completed
16 + 12 * n 32-bit integer  leechers
8 + 12 * N
```
and the code being basically
```
func() {
  writeAction()
  writeTransactionID()
  for torrent in torrents {
    writeSeederLen()
    writeSnatches()
    writeLeechersLen()
  }
}
```